nexus/db-queries/src/db/datastore/deployment.rs

```diff
@@ -410,11 +443,132 @@ impl DataStore {
+            {
+                // Right now, we only implement support for storing SP updates.
+                let PendingMgsUpdateDetails::Sp {
+                    expected_active_version,
+                    expected_inactive_version
+                } = &update.details else {
+                    continue;
+                };
+
+                use nexus_db_schema::schema::hw_baseboard_id::dsl as baseboard_dsl;
+                use nexus_db_schema::schema::bp_pending_mgs_update_sp::dsl as update_dsl;
+                let selection = nexus_db_schema::schema::hw_baseboard_id::table
+                    .select((
+                        DbTypedUuid::from(blueprint_id).into_sql::<diesel::sql_types::Uuid>(),
+                        baseboard_dsl::id,
+                        SpType::from(update.sp_type).into_sql::<SpTypeEnum>(),
+                        // XXX-dap size of int
+                        SpMgsSlot::from(SqlU16::from(u16::try_from(update.slot_id).unwrap())).into_sql::<diesel::sql_types::Int4>(),
+                        ArtifactHash::from(update.artifact_hash).into_sql::<diesel::sql_types::Text>(),
+                        DbArtifactVersion::from(update.artifact_version.clone()).into_sql::<diesel::sql_types::Text>(),
+                        DbArtifactVersion::from(expected_active_version.clone()).into_sql::<diesel::sql_types::Text>(),
+                        match expected_inactive_version {
+                            ExpectedVersion::NoValidVersion => None,
+                            ExpectedVersion::Version(v) => Some(DbArtifactVersion::from(v.clone())),
+                        }
+                        .into_sql::<Nullable<diesel::sql_types::Text>>(),
+                    ))
+                    .filter(
+                        baseboard_dsl::part_number
+                            .eq(update.baseboard_id.part_number.clone()),
+                    )
+                    .filter(
+                        baseboard_dsl::serial_number
+                            .eq(update.baseboard_id.serial_number.clone()),
+                    );
+                let count =
+                   diesel::insert_into(update_dsl::bp_pending_mgs_update_sp)
+                       .values(selection)
+                       .into_columns((
+                           update_dsl::blueprint_id,
+                           update_dsl::hw_baseboard_id,
+                           update_dsl::sp_type,
+                           update_dsl::sp_slot,
+                           update_dsl::artifact_sha256,
+                           update_dsl::artifact_version,
+                           update_dsl::expected_active_version,
+                           update_dsl::expected_inactive_version,
+                       ))
+                       .execute_async(&conn)
+                       .await?;
+                if count != 1 {
+                    // This should be impossible in practice.  We will insert
```

Comment:
Is this what you were referring to in the PR description ("cleanup -- the insertion code is really gross")?